### PR TITLE
Tom geo utils

### PR DESCRIPTION
**Pull request by TomBlock**

### CODE REVIEW

File: snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java [hunk 1]
```diff
@@ -0,0 +1,639 @@
...
+                    0,
+                    rasterDataNode.getRasterWidth(),
+                    rasterDataNode.getRasterHeight());
+        }
+
+        return getGeoBoundary(gc, region, step, usePixelCenter);
+    }
+
+    // @todo 1 tb/tb implement this! 2020-01-31
+//    public static ArrayList<GeoPos[]> createGeoBoundaries(Product product, Rectangle region, int step,
```
Comment (marpet) [inline snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java]: Do we need this method? Isn't it covered by 
`GeneralPath[] createGeoBoundaryPaths(Product product, Rectangle region, int step, final boolean usePixelCenter)`?

File: snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+     * <p>This method is used for an intermediate step when determining a product boundary expressed in geographical
+     * co-ordinates.
+     * <p> This method delegates to {@link #createRectBoundary(java.awt.Rectangle, int, boolean) createRectBoundary(Rectangle, int, boolean)}
+     * and the additional boolean parameter {@code usePixelCenter} is {@code true}.
+     *
+     * @param rect the source rectangle
+     * @param step the mean distance from one pixel position to the other in the returned array
+     * @return the rectangular boundary
+     */
+    public static PixelPos[] createRectBoundary(Rectangle rect, int step) {
```
Comment (marpet) [inline snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java]: Shouldn't this be named createPixelBoundary(...)?

File: snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java [hunk 3]
```diff
@@ -0,0 +1,639 @@
...
+     * This method is used for an intermediate step when determining a product boundary expressed in geographical
+     * co-ordinates.
+     * <p>
+     *
+     * @param rect           the source rectangle
+     * @param step           the mean distance from one pixel position to the other in the returned array
+     * @param usePixelCenter {@code true} if the pixel center should be used
+     * @return the rectangular boundary
+     */
+    static PixelPos[] createRectBoundary(final Rectangle rect, int step, final boolean usePixelCenter) {
```
Comment (marpet) [inline snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java]: Shouldn't this be named createPixelBoundary(...)?

File: snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java [hunk 4]
```diff
@@ -0,0 +1,639 @@
...
+        int xStart = rect.x;
+        int yStart = rect.y;
+        final int w = usePixelCenter ? rect.width - 1 : rect.width;
+        final int h = usePixelCenter ? rect.height - 1 : rect.height;
+        int xEnd = xStart + w;
+        int yEnd = yStart + h;
+
+        final ArrayList<GeoPos> geoPosList = new ArrayList<>();
+
+        final int validYMin = getFirstValid_Y(gc, insetDistance, xStart, yStart, xEnd, yEnd);
```
Comment (marpet) [inline snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java]: Could it make sense to parallelise the computation? Maybe using an ExecutorService?

File: snap-core/src/main/java/org/esa/snap/core/util/ProductUtils.java [hunk 5]
```diff
@@ -574,74 +503,47 @@ private static void checkCanceled(ProgressMonitor pm) throws IOException {
...
-     * @param region         the region rectangle in product pixel coordinates, can be null for entire product
-     * @param step           the step given in pixels
-     * @param usePixelCenter {@code true} if the pixel center should be used to create the pathes
-     * @return an array of shape objects
-     *
-     * @throws IllegalArgumentException if product is null or if the product's {@link GeoCoding} is null
-     * @see #createGeoBoundary(Product, java.awt.Rectangle, int, boolean)
+     * @deprecated use {@link GeoUtils} instead
      */
+    @Deprecated
```
Comment (marpet) [inline snap-core/src/main/java/org/esa/snap/core/util/ProductUtils.java]: Add also 'use {@link GeoUtils} instead'?

File: snap-core/src/main/java/org/esa/snap/core/util/ProductUtils.java [hunk 6]
```diff
@@ -657,6 +559,10 @@ private static void checkCanceled(ProgressMonitor pm) throws IOException {
         return pathList.toArray(new GeneralPath[pathList.size()]);
     }
 
+    /**
+     * @deprecated use {@link GeoUtils} instead
```
Comment (marpet) [inline snap-core/src/main/java/org/esa/snap/core/util/ProductUtils.java]: Shouldn't the deprecated methods make use os the new GeoUtils methods?

File: snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java (hunk 1 above)
Comment (TomBlock) [reply snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java]: This is a planned API method to allow for segmented geo-bounds, i.e. geo-boundaries that consist of multiple segments separated by fill-value regions. We should discuss this separately, that is a non-trivial task to integrate this into SNAP

File: snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java (hunk 2 above)
Comment (TomBlock) [reply snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java]: better: createPixelBoundaryFromRect()

what do you think?

File: snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java (hunk 3 above)
Comment (TomBlock) [reply snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java]: same, see above

File: snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java (hunk 4 above)
Comment (TomBlock) [reply snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java]: Absolutely

File: snap-core/src/main/java/org/esa/snap/core/util/ProductUtils.java (hunk 5 above)
Comment (TomBlock) [reply snap-core/src/main/java/org/esa/snap/core/util/ProductUtils.java]: I completed this by @deprecated no replacement

File: snap-core/src/main/java/org/esa/snap/core/util/ProductUtils.java (hunk 6 above)
Comment (TomBlock) [reply snap-core/src/main/java/org/esa/snap/core/util/ProductUtils.java]: done

File: snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java (hunk 2 above)
Comment (marpet) [reply snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java]: Agreed!

File: snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java (hunk 1 above)
Comment (marpet) [reply snap-core/src/main/java/org/esa/snap/core/util/GeoUtils.java]: Sure, okay!

**Pull request merged**